packages/compass-query-bar/src/components/query-bar/query-bar.tsx

```diff
@@ -49,74 +61,260 @@ const openQueryHistoryStyles = cx(
   focusRingStyles
 );
 
+const rowStyles = css({
+  alignItems: 'center',
+  display: 'flex',
+  flexGrow: 1,
+  position: 'relative',
+  margin: spacing[1],
+  padding: `0 ${spacing[2]}px`,
+  gap: spacing[3],
+});
+
+const firstQueryOptionsRowStyles = css({
+  margin: `${spacing[1]}px 0px`,
+  padding: 0,
+});
+
+const queryDocsLinkStyles = css({
+  flexShrink: 0,
+  marginRight: spacing[1],
+});
+
+type QueryBarOptionProps = {
+  filterPlaceholder?: string; // The placeholder text for the filter input.
+  filterValid: boolean; // Whether the filter is valid.
+  filterString: string; // The value of the `filter`.
+
+  projectPlaceholder?: string;
+  projectValid: boolean;
+  projectString: string;
+
+  sortPlaceholder?: string;
+  sortValid: boolean;
+  sortString: string;
+
+  collationPlaceholder?: string;
+  collationValid: boolean;
+  collationString: string;
+
+  skipPlaceholder?: string;
+  skipValid: boolean;
+  skipString: string;
+
+  limitPlaceholder?: string;
+  limitValid: boolean;
+  limitString: string;
+
+  maxTimeMSPlaceholder?: string;
+  maxTimeMSValid: boolean;
+  maxTimeMSString: string;
+};
+
 type QueryBarProps = {
+  autoPopulated: boolean;
   buttonLabel?: string;
   expanded: boolean;
-  isQueryOptionsExpanded?: boolean;
-  valid: boolean;
+  layout?: QueryBarLayout;
+  onApply: () => void;
+  onChangeQueryOption: (queryOption: QueryOption, value: string) => void;
+  onReset: () => void;
   queryState: 'apply' | 'reset';
+  refreshEditorAction: Listenable;
+  schemaFields: string[];
+  serverVersion: string;
   showQueryHistoryButton?: boolean;
   toggleExpandQueryOptions: () => void;
   toggleQueryHistory: () => void;
-};
+  valid: boolean;
+} & QueryBarOptionProps;
 
 export const QueryBar: React.FunctionComponent<QueryBarProps> = ({
+  autoPopulated,
   buttonLabel = 'Apply',
   expanded: isQueryOptionsExpanded = false,
-  valid: isQueryValid,
+  layout = [
+    'filter',
+    ['project', 'sort'],
+    ['collation', 'skip', 'limit', 'maxTimeMS'],
+  ],
```

Comment:
Is there ever a case where `filter` is not the first and only option in the first row? Seems like we are complicating both interface and implementation (both component logic and styles) to support that, but it doesn't look like this is really needed?

Reply:
Yeah, I checked mms layouts and there `filter` is always first and single row, do we really have the use-case for having it so flexible?

If you don't want to do this in this scope I won't try to convince you otherwise, FWIW if the goal is to "simplify", moving existing code to typescript without re-evaluating the purpose of it is just covering up possible existing issues